server/routes/api/users.js

```diff
@@ -0,0 +1,101 @@
+const express = require("express");
+const jwt = require("jsonwebtoken");
+const bcrypt = require("bcrypt");
+
+const User = require("../../models/User");
+const { passportSecret } = require("../../config/keys");
+const validateEmail = require("../../validation/email");
+const validatePassword = require("../../validation/password");
+
+const router = express.Router();
+
+const validateSignupInputs = (body) => {
+  const { name, email, password } = body;
+  let isValid = true;
+  if (!name || !validateEmail(email) || !validatePassword(password))
+    isValid = false;
+
+  return isValid;
+};
+
+router.post("/signup", async (req, res) => {
+  if (!validateSignupInputs(req.body)) {
+    res.status(400).send({
+      response: "Invalid input.",
+    });
+    return;
+  }
+
+  const { name, email, password } = req.body;
+
+  try {
+    // check if email exists in database
+    const user = await User.findOne({ email });
+    if (user) {
+      res.status(409).send({
+        response: "An account already exists with that email address.",
+      });
+      return;
+    }
+
+    // register user
+    const newUser = new User({
+      name,
+      email,
+      password,
+    });
+    await newUser.save(); // password hashed on save
+
+    // sign and return jwt as Bearer token in Authorization header
+    const payload = {
+      id: newUser.id,
+      name: newUser.name,
+      email: newUser.email,
+    };
+    const token = jwt.sign(payload, passportSecret);
+    res.status(201).send({ token });
+  } catch (err) {
+    // eslint-disable-next-line no-console
+    console.log(err);
+    res.status(500).end();
```

Comment:
I don't usually like sending 500s as a catch all (because that would happen anyways). Really we should only be trying and catching if we know the error. But I think this is fine - just something to note. 
Probably want to also not use console.log here, but console.error or whatever is used to log errors